core/dbt/cli/flags.py

```diff
@@ -277,3 +279,106 @@ def fire_deprecations(self):
         # It is necessary to remove this attr from the class so it does
         # not get pickled when written to disk as json.
         object.__delattr__(self, "deprecated_env_var_warnings")
+
+    @classmethod
+    def from_dict(cls, cmd: CliCommand, d: Dict[str, Any]) -> "Flags":
+        arg_list = get_args_for_cmd_from_dict(cmd, d)
+        ctx = args_to_context(arg_list)
+        flags = cls(ctx=ctx)
+        flags.fire_deprecations()
+        return flags
+
+
+def get_args_for_cmd_from_dict(cmd: CliCommand, d: Dict[str, Any]) -> List[str]:
+    """Given a command name and a dict, returns a list of strings representing
+    the CLI arguments that for a command. The order of this list is consistent with
+    which flags are expected at the parent level vs the command level.
+
+    e.g. fn("run", {"defer": True, "print": False}) -> ["--no-print", "run", "--defer"]
+
+    The result of this function can be passed in to the args_to_context function
+    to produce a click context to instantiate Flags with.
+    """
+
+    cmd_args = get_args_for_cmd(cmd)
+    parent_args = get_parent_args()
+    default_args = [x.lower() for x in FLAGS_DEFAULTS.keys()]
+
+    res = cmd.to_list()
+
+    for k, v in d.items():
+        k = k.lower()
+
+        # if a "which" value exists in the args dict, it should match the cmd arg
+        if k == "which":
+            if v != cmd.value:
+                raise DbtInternalError(f"cmd '{cmd.value}' does not match value of which: '{v}'")
+            continue
+
+        # param was assigned from defaults and should not be included
+        if k not in cmd_args + parent_args and k in default_args:
+            continue
+
+        # if the param is in parent args, it should come before the arg name
+        # e.g. ["--print", "run"] vs ["run", "--print"]
+        add_fn = res.append
+        if k in parent_args:
+
+            def add_fn(x):
+                res.insert(0, x)
+
+        spinal_cased = k.replace("_", "-")
+
+        if v in (None, False):
+            add_fn(f"--no-{spinal_cased}")
+        elif v is True:
+            add_fn(f"--{spinal_cased}")
+        else:
+            add_fn(f"--{spinal_cased}={v}")
+
+    return res
+
+
+def get_parent_args() -> List[str]:
+    """Return a list representing the params the base click command takes."""
+    from dbt.cli.main import cli
+
+    return format_cmd_params(cli.params)
+
+
+def get_args_for_cmd(cmd: CliCommand) -> List[str]:
+    """Given a command, return a list of strings representing the params
+    that command takes. This function only returns params assigned to a
+    specific command, not those of its parent command.
+
+    e.g. fn("run") -> ["defer", "favor_state", "exclude", ...]
+    """
+    import dbt.cli.main as cli
+
+    CMD_DICT: Dict[CliCommand, ClickCommand] = {
```

Comment:
I'd rather have an exhaustive list of click commands mapped to an enum than use string matching.

Reply:
Can we move this to Definition at the top of the file? Or there's circular dependency so you can't do it?

Reply:
Also how do we detect this list get out of sync with our click command definition? Can we add some unittest to catch that?

Reply:
This dict is defined here to prevent a cyclical dependency.

Even if we didn't use an enum and instead used string matching for commands, we would still need to update some sort of list or validation somewhere when adding new commands which would become stale over time (e.g. does `source_freshness` translate to `cli.main.freshness` or only the string `freshness`?). Using an enum means you have one entry to add in an enum and one in a dict.

I will add a section to `cli/README.md` about the steps to take when adding a new command, but I'm not sure what unit tests we could add to make sure a contributor updates this dict or the enum of the commands.

Reply:
Maybe something like looping over all of the click command and see if it is in the values of this dict? My main concern is people will forget about it even if it is in README.